In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

In [3]:
df = pd.read_csv("../data/raw/electronics_inventory.csv")

df.head()

,Date,Product_ID,Product_Name,Category,Brand,Current_Stock,Units_Sold,Unit_Price,Discount_Percent,Promotion,Holiday,Lead_Time_Days,Safety_Stock,Stockout
0,2022-02-17,P001,iPhone 13,Mobile,Apple,154.0,9,65000.0,0.0,0.0,0,7.0,15,0
1,2026-06-06,P008,Samsung 55-inch Smart TV,TV,Samsung,101.0,3,120000.0,0.0,1.0,0,14.0,5,0
2,2025-06-04,P006,HP Pavilion 15,Laptop,HP,101.0,5,95000.0,0.0,1.0,0,9.0,8,0
3,2022-11-14,P004,Redmi Note 13,Mobile,Xiaomi,91.0,10,32000.0,0.0,0.0,0,5.0,20,0
4,2025-08-08,P001,iPhone 13,Mobile,Apple,129.0,5,65000.0,0.0,0.0,0,7.0,15,0


In [4]:
df.columns

Index(['Date', 'Product_ID', 'Product_Name', 'Category', 'Brand',
       'Current_Stock', 'Units_Sold', 'Unit_Price', 'Discount_Percent',
       'Promotion', 'Holiday', 'Lead_Time_Days', 'Safety_Stock', 'Stockout'],
      dtype='object')

In [6]:
df.shape

(16815, 14)

In [7]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16815 entries, 0 to 16814
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date              16815 non-null  object 
 1   Product_ID        16815 non-null  object 
 2   Product_Name      16815 non-null  object 
 3   Category          16815 non-null  object 
 4   Brand             16815 non-null  object 
 5   Current_Stock     16749 non-null  float64
 6   Units_Sold        16815 non-null  int64  
 7   Unit_Price        16749 non-null  float64
 8   Discount_Percent  16681 non-null  float64
 9   Promotion         16765 non-null  float64
 10  Holiday           16815 non-null  int64  
 11  Lead_Time_Days    16765 non-null  float64
 12  Safety_Stock      16815 non-null  int64  
 13  Stockout          16815 non-null  int64  
dtypes: float64(5), int64(4), object(5)
memory usage: 1.8+ MB


In [8]:
df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Date,16815,1674,2024-01-10,12,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Product_ID,16815,10,P001,1690,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Product_Name,16815,10,iPhone 13,1690,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Category,16815,3,Mobile,6729,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Brand,16815,8,Apple,3372,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Current_Stock,16749.0,NaN,NaN,NaN,107.325393,31.18799,0.0,87.0,106.0,128.0,230.0
Units_Sold,16815.0,NaN,NaN,NaN,4.934463,3.953156,0.0,2.0,4.0,7.0,110.0
Unit_Price,16749.0,NaN,NaN,NaN,80509.522957,24824.25417,32000.0,65000.0,78000.0,95000.0,120000.0
Discount_Percent,16681.0,NaN,NaN,NaN,0.281758,1.943842,0.0,0.0,0.0,0.0,20.0
Promotion,16765.0,NaN,NaN,NaN,0.157232,0.364031,0.0,0.0,0.0,0.0,1.0


In [11]:
print(df.isnull().sum())

Date                  0
Product_ID            0
Product_Name          0
Category              0
Brand                 0
Current_Stock        66
Units_Sold            0
Unit_Price           66
Discount_Percent    134
Promotion            50
Holiday               0
Lead_Time_Days       50
Safety_Stock          0
Stockout              0
dtype: int64


| Column             | Missing |
| ------------------ | ------: |
| `Current_Stock`    |      66 |
| `Unit_Price`       |      66 |
| `Discount_Percent` |     134 |
| `Promotion`        |      50 |


In [12]:
print("Duplicate Rows:", df.duplicated().sum())

Duplicate Rows: 75


In [13]:
print("Date Range:")
print("Start:", df["Date"].min())
print("End:", df["Date"].max())

Date Range:
Start: 2022-01-01
End: 2026-08-01


In [14]:
print("Number of Products:", df["Product_ID"].nunique())

Number of Products: 10


In [15]:
print("Products:")
print(df["Product_Name"].unique())

Products:
['iPhone 13' 'Samsung 55-inch Smart TV' 'HP Pavilion 15' 'Redmi Note 13'
 'iPhone 15' 'Samsung Galaxy A55' 'Dell Inspiron 15' 'Lenovo IdeaPad 3'
 'Sony 50-inch Bravia TV' 'LG 43-inch Smart TV']


In [16]:
print("Categories:")
print(df["Category"].value_counts())

Categories:
Category
Mobile    6729
TV        5047
Laptop    5039
Name: count, dtype: int64


In [17]:
product_counts = df.groupby(
    ["Product_ID", "Product_Name", "Category"]
).size().reset_index(name="Record_Count")

product_counts

,Product_ID,Product_Name,Category,Record_Count
0,P001,iPhone 13,Mobile,1690
1,P002,iPhone 15,Mobile,1682
2,P003,Samsung Galaxy A55,Mobile,1683
3,P004,Redmi Note 13,Mobile,1674
4,P005,Dell Inspiron 15,Laptop,1678
5,P006,HP Pavilion 15,Laptop,1679
6,P007,Lenovo IdeaPad 3,Laptop,1682
7,P008,Samsung 55-inch Smart TV,TV,1682
8,P009,LG 43-inch Smart TV,TV,1684
9,P010,Sony 50-inch Bravia TV,TV,1681


In [18]:
product_counts.sort_values("Record_Count")

,Product_ID,Product_Name,Category,Record_Count
3,P004,Redmi Note 13,Mobile,1674
4,P005,Dell Inspiron 15,Laptop,1678
5,P006,HP Pavilion 15,Laptop,1679
9,P010,Sony 50-inch Bravia TV,TV,1681
6,P007,Lenovo IdeaPad 3,Laptop,1682
7,P008,Samsung 55-inch Smart TV,TV,1682
1,P002,iPhone 15,Mobile,1682
2,P003,Samsung Galaxy A55,Mobile,1683
8,P009,LG 43-inch Smart TV,TV,1684
0,P001,iPhone 13,Mobile,1690


In [19]:
df["Date"] = pd.to_datetime(df["Date"])

yearly_product_counts = (
    df.groupby([df["Date"].dt.year, "Product_Name"])
      .size()
      .unstack(fill_value=0)
)

yearly_product_counts

Product_Name,Dell Inspiron 15,HP Pavilion 15,LG 43-inch Smart TV,Lenovo IdeaPad 3,Redmi Note 13,Samsung 55-inch Smart TV,Samsung Galaxy A55,Sony 50-inch Bravia TV,iPhone 13,iPhone 15
Date,,,,,,,,,,
2022,365,367,369,368,365,366,367,367,370,367
2023,365,367,366,366,365,366,367,366,369,367
2024,370,367,370,367,366,368,368,367,370,368
2025,365,365,366,366,365,367,367,368,367,365
2026,213,213,213,215,213,215,214,213,214,215


In [20]:
df[
    [
        "Current_Stock",
        "Units_Sold",
        "Unit_Price",
        "Discount_Percent",
        "Lead_Time_Days",
        "Safety_Stock"
    ]
].describe().T

,count,mean,std,min,25%,50%,75%,max
Current_Stock,16749.0,107.325393,31.187990,0.0,87.0,106.0,128.0,230.0
Units_Sold,16815.0,4.934463,3.953156,0.0,2.0,4.0,7.0,110.0
Unit_Price,16749.0,80509.522957,24824.254170,32000.0,65000.0,78000.0,95000.0,120000.0
Discount_Percent,16681.0,0.281758,1.943842,0.0,0.0,0.0,0.0,20.0
Lead_Time_Days,16765.0,9.401372,3.039745,5.0,7.0,10.0,12.0,14.0
Safety_Stock,16815.0,10.698959,5.402675,5.0,6.0,8.0,15.0,20.0


In [21]:
for col in ["Category", "Brand", "Promotion", "Holiday", "Stockout"]:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False))


--- Category ---
Category
Mobile    6729
TV        5047
Laptop    5039
Name: count, dtype: int64

--- Brand ---
Brand
Apple      3372
Samsung    3365
LG         1684
Lenovo     1682
Sony       1681
HP         1679
Dell       1678
Xiaomi     1674
Name: count, dtype: int64

--- Promotion ---
Promotion
0.0    14129
1.0     2636
NaN       50
Name: count, dtype: int64

--- Holiday ---
Holiday
0    16635
1      180
Name: count, dtype: int64

--- Stockout ---
Stockout
0    16559
1      256
Name: count, dtype: int64


In [22]:
df.nlargest(20, "Units_Sold")[
    [
        "Date",
        "Product_ID",
        "Product_Name",
        "Category",
        "Current_Stock",
        "Units_Sold",
        "Unit_Price",
        "Discount_Percent",
        "Promotion",
        "Holiday",
        "Lead_Time_Days",
        "Safety_Stock",
        "Stockout"
    ]
]

,Date,Product_ID,Product_Name,Category,Current_Stock,Units_Sold,Unit_Price,Discount_Percent,Promotion,Holiday,Lead_Time_Days,Safety_Stock,Stockout
2799,2022-05-14,P003,Samsung Galaxy A55,Mobile,132.0,110,52000.0,0.0,0.0,0,6.0,18,0
12531,2024-12-05,P004,Redmi Note 13,Mobile,107.0,110,32000.0,0.0,0.0,0,5.0,20,0
15556,2025-03-22,P003,Samsung Galaxy A55,Mobile,160.0,110,52000.0,0.0,0.0,0,6.0,18,0
9505,2026-01-05,P001,iPhone 13,Mobile,138.0,88,65000.0,0.0,0.0,0,7.0,15,0
7628,2022-08-26,P004,Redmi Note 13,Mobile,88.0,72,32000.0,0.0,0.0,0,5.0,20,0
1528,2025-12-14,P001,iPhone 13,Mobile,161.0,66,65000.0,0.0,0.0,0,7.0,15,0
1077,2022-08-03,P004,Redmi Note 13,Mobile,89.0,63,32000.0,0.0,0.0,0,5.0,20,0
5147,2024-08-20,P002,iPhone 15,Mobile,114.0,56,85000.0,0.0,0.0,0,7.0,15,0
15944,2024-03-09,P002,iPhone 15,Mobile,176.0,55,85000.0,0.0,1.0,0,7.0,15,0
13273,2025-10-26,P004,Redmi Note 13,Mobile,140.0,50,32000.0,0.0,0.0,0,5.0,20,0


In [23]:
df.nsmallest(20, "Units_Sold")[
    [
        "Date",
        "Product_Name",
        "Current_Stock",
        "Units_Sold",
        "Promotion",
        "Holiday",
        "Stockout"
    ]
]

,Date,Product_Name,Current_Stock,Units_Sold,Promotion,Holiday,Stockout
17,2025-11-03,Lenovo IdeaPad 3,126.0,0,0.0,0,0
25,2025-10-09,Samsung 55-inch Smart TV,69.0,0,0.0,0,0
28,2024-12-25,Sony 50-inch Bravia TV,68.0,0,0.0,0,0
34,2024-06-28,iPhone 15,0.0,0,0.0,0,1
40,2024-12-25,Lenovo IdeaPad 3,80.0,0,0.0,0,0
47,2026-01-30,LG 43-inch Smart TV,0.0,0,0.0,0,1
65,2023-06-14,Sony 50-inch Bravia TV,64.0,0,0.0,0,0
68,2025-01-04,Sony 50-inch Bravia TV,95.0,0,0.0,0,0
109,2022-01-08,Dell Inspiron 15,81.0,0,0.0,0,0
153,2022-02-03,Sony 50-inch Bravia TV,46.0,0,0.0,0,0


In [24]:
df[df["Units_Sold"] >= 50][
    [
        "Date",
        "Product_ID",
        "Product_Name",
        "Units_Sold",
        "Current_Stock",
        "Discount_Percent",
        "Promotion",
        "Holiday",
        "Stockout"
    ]
].sort_values("Units_Sold", ascending=False)

,Date,Product_ID,Product_Name,Units_Sold,Current_Stock,Discount_Percent,Promotion,Holiday,Stockout
12531,2024-12-05,P004,Redmi Note 13,110,107.0,0.0,0.0,0,0
2799,2022-05-14,P003,Samsung Galaxy A55,110,132.0,0.0,0.0,0,0
15556,2025-03-22,P003,Samsung Galaxy A55,110,160.0,0.0,0.0,0,0
9505,2026-01-05,P001,iPhone 13,88,138.0,0.0,0.0,0,0
7628,2022-08-26,P004,Redmi Note 13,72,88.0,0.0,0.0,0,0
1528,2025-12-14,P001,iPhone 13,66,161.0,0.0,0.0,0,0
1077,2022-08-03,P004,Redmi Note 13,63,89.0,0.0,0.0,0,0
5147,2024-08-20,P002,iPhone 15,56,114.0,0.0,0.0,0,0
15944,2024-03-09,P002,iPhone 15,55,176.0,0.0,1.0,0,0
13273,2025-10-26,P004,Redmi Note 13,50,140.0,0.0,0.0,0,0


In [25]:
df.groupby("Product_Name")["Units_Sold"].agg(
    ["count", "mean", "median", "std", "min", "max"]
).sort_values("max", ascending=False)


,count,mean,median,std,min,max
Product_Name,,,,,,
Samsung Galaxy A55,1683,8.359477,8.0,4.677692,0,110
Redmi Note 13,1674,8.350060,8.0,4.613376,0,110
iPhone 13,1690,7.702959,7.0,3.844897,0,88
iPhone 15,1682,6.847206,7.0,3.500126,0,56
Samsung 55-inch Smart TV,1682,2.666468,2.0,2.297334,0,45
Lenovo IdeaPad 3,1682,3.547562,3.0,2.272183,0,44
HP Pavilion 15,1679,4.455628,4.0,2.347423,0,35
Dell Inspiron 15,1678,3.184744,3.0,1.955792,0,33
LG 43-inch Smart TV,1684,2.562945,2.0,1.756125,0,20
